mk/checkconf.mk

```diff
@@ -21,6 +21,7 @@ define check-conf-h
 	mkdir -p $(dir $@);					\
 	echo "#ifndef $${guard}" >$@.tmp;			\
 	echo "#define $${guard}" >>$@.tmp;			\
+	[ "${CFG_TEE_CORE_DEBUG}" != "y" ] && echo "#define NDEBUG" >>$@.tmp; \
```

Comment:
I'd prefer that you handle the link between `CFG_TEE_CORE_DEBUG` and `NDEBUG` in `mk/config.mk`:
```
CFG_TEE_CORE_DEBUG ?= n
ifneq ($(CFG_TEE_CORE_DEBUG),y)
NDEBUG ?= 1
endif
```
Rationale: NDEBUG becomes a full-fledged configuration variable (like the `CFG_` ones). That is, it can be set in the makefiles, or on the command line, via environment variables etc. with the usual priorities (as documented at the beginning of `mk/config.mk`).

Then instead of adding special cases to `check-conf-h` and `check-conf-mk`, you simply update the list of variables we export to the `conf.h` and `conf.mk` (that's in `cfg-vars-by-prefix` and in `check-conf-mk` [1]).
For clarity, here's a diff with your proposed patch...
```
diff --git a/mk/checkconf.mk b/mk/checkconf.mk
index 36b62c3..444edaf 100644
--- a/mk/checkconf.mk
+++ b/mk/checkconf.mk
@@ -21,7 +21,6 @@ define check-conf-h
        mkdir -p $(dir $@);                                     \
        echo "#ifndef $${guard}" >$@.tmp;                       \
        echo "#define $${guard}" >>$@.tmp;                      \
-       [ "${CFG_TEE_CORE_DEBUG}" != "y" ] && echo "#define NDEBUG" >>$@.tmp; \
        echo -n "$${cnf}" | sed 's/_nl_ */\n/g' >>$@.tmp;       \
        echo "#endif" >>$@.tmp;                                 \
        $(call mv-if-changed,$@.tmp,$@)
@@ -31,7 +30,7 @@ define check-conf-mk
        $(q)set -e;                                             \
        $(cmd-echo-silent) '  CHK     $@';                      \
        cnf="$(strip $(foreach var,                             \
-               $(call cfg-vars-by-prefix,CFG_),                \
+               $(call cfg-vars-by-prefix,CFG_ NDEBUG),         \
                $(strip $(var)=$($(var))_nl_)))"; \
        mkdir -p $(dir $@);                                     \
        echo "# auto-generated TEE configuration file" >$@.tmp; \
@@ -39,7 +38,6 @@ define check-conf-mk
        echo "ARCH=${ARCH}" >>$@.tmp;                           \
        echo "PLATFORM=${PLATFORM}" >>$@.tmp;                   \
        echo "PLATFORM_FLAVOR=${PLATFORM_FLAVOR}" >>$@.tmp;     \
-       [ "${CFG_TEE_CORE_DEBUG}" != "y" ] && echo "NDEBUG=1" >>$@.tmp; \
        echo -n "$${cnf}" | sed 's/_nl_ */\n/g' >>$@.tmp;       \
        $(call mv-if-changed,$@.tmp,$@)
 endef
@@ -56,7 +54,7 @@ endef
 
 define cfg-vars-by-prefix
        $(strip $(if $(1),$(call _cfg-vars-by-prefix,$(1)),
-                         $(call _cfg-vars-by-prefix,CFG_ _CFG_ PLATFORM_)))
+                         $(call _cfg-vars-by-prefix,CFG_ _CFG_ PLATFORM_ NDEBUG)))
 endef
 
 define _cfg-vars-by-prefix
diff --git a/mk/config.mk b/mk/config.mk
index d302f55..18e1175 100644
--- a/mk/config.mk
+++ b/mk/config.mk
@@ -41,6 +41,9 @@ WARNS ?= 3
 # If y, enable debug mode of the tee firmware (CPU restart, Core Status
 # verbose, panic & assert verbose). When disable, NDEBUG directive is defined.
 CFG_TEE_CORE_DEBUG ?= n
+ifneq ($(CFG_TEE_CORE_DEBUG),y)
+NDEBUG ?= 1
+endif
 
 # Max level of the tee core traces. 0 means disable, 4 is max.
 # Supported values: 0 (no traces) to 4 (all traces)
```

It seems to work as expected:
```
$ make -j9 -s
$ grep -e CFG_TEE_CORE_DEBUG -e NDEBUG out/arm-plat-vexpress/conf.mk out/arm-plat-vexpress/include/generated/conf.h 
out/arm-plat-vexpress/conf.mk:CFG_TEE_CORE_DEBUG=n
out/arm-plat-vexpress/conf.mk:NDEBUG=1
out/arm-plat-vexpress/include/generated/conf.h:/* CFG_TEE_CORE_DEBUG is not set ('n') */
out/arm-plat-vexpress/include/generated/conf.h:#define NDEBUG 1 /* '1' */

$ make -j9 -s CFG_TEE_CORE_DEBUG=y
$ grep -e CFG_TEE_CORE_DEBUG -e NDEBUG out/arm-plat-vexpress/conf.mk out/arm-plat-vexpress/include/generated/conf.h 
out/arm-plat-vexpress/conf.mk:CFG_TEE_CORE_DEBUG=y
out/arm-plat-vexpress/include/generated/conf.h:#define CFG_TEE_CORE_DEBUG 1 /* 'y' */

$  make -j9 -s NDEBUG=
$ grep -e CFG_TEE_CORE_DEBUG -e NDEBUG out/arm-plat-vexpress/conf.mk out/arm-plat-vexpress/include/generated/conf.h 
out/arm-plat-vexpress/conf.mk:CFG_TEE_CORE_DEBUG=n
out/arm-plat-vexpress/conf.mk:NDEBUG=
out/arm-plat-vexpress/include/generated/conf.h:/* CFG_TEE_CORE_DEBUG is not set ('n') */
out/arm-plat-vexpress/include/generated/conf.h:/* NDEBUG is not set ('') */

$ make -j9 -s CFG_TEE_CORE_DEBUG=y NDEBUG=1
$ grep -e CFG_TEE_CORE_DEBUG -e NDEBUG out/arm-plat-vexpress/conf.mk out/arm-plat-vexpress/include/generated/conf.h 
out/arm-plat-vexpress/conf.mk:CFG_TEE_CORE_DEBUG=y
out/arm-plat-vexpress/conf.mk:NDEBUG=1
out/arm-plat-vexpress/include/generated/conf.h:#define CFG_TEE_CORE_DEBUG 1 /* 'y' */
out/arm-plat-vexpress/include/generated/conf.h:#define NDEBUG 1 /* '1' */
```

[1] I would also prefer if the list of variables exported to `conf.h` were in `check-conf-h`, not in the default args of `cfg-vars-by-prefix`. Let's keep that for another commit.